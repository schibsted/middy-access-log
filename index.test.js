const middy = require('@middy/core');
const createError = require('http-errors');
const middleware = require('./index');

jest.mock('perf_hooks', () => {
    const performanceMock = {
        now: jest.fn().mockReturnValue(1),
    };
    return { performance: performanceMock };
});

test('Middleware logs on success', async () => {
    const handler = middy(async () => ({
        statusCode: 200,
        body: JSON.stringify({ foo: 'bar' }),
        headers: {
            someHeader: 'someValue',
        },
    }));

    const mockLogger = {
        info: jest.fn(() => {}),
    };

    handler.use(middleware({ logger: mockLogger }));

    const event = {
        headers: {
            host: 'localhost:3000',
            connection: 'keep-alive',
            'upgrade-insecure-requests': '1',
            'user-agent':
                'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36',
            accept:
                'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-encoding': 'gzip, deflate, br',
            'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5',
            cookie: 'my cookie 1; my cookie 2',
        },
        path: '/foobar',
        httpMethod: 'GET',
    };

    await handler(event, {});
    expect(mockLogger.info).toHaveBeenCalledTimes(1);
    expect(mockLogger.info.mock.calls[0]).toMatchSnapshot();
});

test('Middleware logs on error', async () => {
    const error = new createError.InternalServerError('whoops');
    const handler = middy(async () => {
        throw error;
    });

    const mockLogger = {
        info: jest.fn(() => {}),
    };

    handler.use(middleware({ logger: mockLogger }));

    const event = {
        headers: {
            host: 'localhost:3000',
            connection: 'keep-alive',
            'upgrade-insecure-requests': '1',
            'user-agent':
                'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36',
            accept:
                'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-encoding': 'gzip, deflate, br',
            'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5',
            cookie: 'my cookie 1; my cookie 2',
        },
        path: '/foobar',
        httpMethod: 'GET',
    };

    await expect(handler(event, {})).rejects.toEqual(error);

    expect(mockLogger.info).toHaveBeenCalledTimes(1);
    expect(mockLogger.info.mock.calls[0]).toMatchSnapshot();
});

test('Middleware skips sensitive headers on success', async () => {
    const handler = middy(async () => ({
        statusCode: 200,
        body: JSON.stringify({ foo: 'bar' }),
        headers: {
            someHeader: 'someValue',
        },
    }));

    const mockLogger = {
        info: jest.fn(() => {}),
    };

    handler.use(
        middleware({ logger: mockLogger, excludeHeaders: ['cookie', 'authentication', 'customRequestSignature'] })
    );

    const event = {
        headers: {
            host: 'localhost:3000',
            connection: 'keep-alive',
            'upgrade-insecure-requests': '1',
            'user-agent':
                'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36',
            accept:
                'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-encoding': 'gzip, deflate, br',
            'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5',
            cookie: 'my cookie 1; my cookie 2',
            authentication: 'super secret token',
            customRequestSignature: 'foobar 123',
        },
        path: '/foobar',
        httpMethod: 'GET',
    };

    await handler(event, {});
    expect(mockLogger.info).toHaveBeenCalledTimes(1);
    expect(mockLogger.info.mock.calls[0]).toMatchSnapshot();
});

test('Middleware skips sensitive headers on error', async () => {
    const error = new createError.InternalServerError('whoops');
    const handler = middy(async () => {
        throw error;
    });

    const mockLogger = {
        info: jest.fn(() => {}),
    };

    handler.use(
        middleware({ logger: mockLogger, excludeHeaders: ['cookie', 'authentication', 'customRequestSignature'] })
    );
    const event = {
        headers: {
            host: 'localhost:3000',
            connection: 'keep-alive',
            'upgrade-insecure-requests': '1',
            'user-agent':
                'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36',
            accept:
                'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9',
            'accept-encoding': 'gzip, deflate, br',
            'accept-language': 'en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5',
            cookie: 'my cookie 1; my cookie 2',
            authentication: 'super secret token',
            customRequestSignature: 'foobar 123',
        },
        path: '/foobar',
        httpMethod: 'GET',
    };

    await expect(handler(event, {})).rejects.toEqual(error);

    expect(mockLogger.info).toHaveBeenCalledTimes(1);
    expect(mockLogger.info.mock.calls[0]).toMatchSnapshot();
});
