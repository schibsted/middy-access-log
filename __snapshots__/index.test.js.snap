// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Middleware logs on error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "duration": 0,
        "req": Object {
          "headers": Object {
            "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
            "accept-encoding": "gzip, deflate, br",
            "accept-language": "en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5",
            "connection": "keep-alive",
            "host": "localhost:3000",
            "upgrade-insecure-requests": "1",
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
          },
          "method": "GET",
          "url": "/foobar",
        },
        "res": Object {},
      },
      "  --> GET /foobar  0ns",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Middleware logs on success 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "duration": 0,
        "req": Object {
          "headers": Object {
            "accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9",
            "accept-encoding": "gzip, deflate, br",
            "accept-language": "en-GB,en;q=0.9,en-US;q=0.8,pl;q=0.7,nb;q=0.6,no;q=0.5",
            "connection": "keep-alive",
            "host": "localhost:3000",
            "upgrade-insecure-requests": "1",
            "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.111 Safari/537.36",
          },
          "method": "GET",
          "url": "/foobar",
        },
        "res": Object {
          "headers": Object {
            "someHeader": "someValue",
          },
          "statusCode": 200,
        },
      },
      "  --> GET /foobar 200 0ns",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
